samplers/aws/xray/internal/rule.go

```diff
@@ -0,0 +1,214 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal // import "go.opentelemetry.io/contrib/samplers/aws/xray/internal"
+
+import (
+	"sync/atomic"
+	"time"
+
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// Rule represents a sampling rule which contains rule properties and reservoir which keeps tracks of sampling statistics of a rule
+type Rule struct {
+	samplingStatistics *samplingStatistics
+
+	// reservoir has equivalent fields to store what we receive from service API getSamplingTargets
+	// https://docs.aws.amazon.com/xray/latest/api/API_GetSamplingTargets.html
+	reservoir reservoir
+
+	// equivalent to what we receive from service API getSamplingRules
+	// https://docs.aws.amazon.com/cli/latest/reference/xray/get-sampling-rules.html
+	ruleProperties ruleProperties
+}
+
+type samplingStatistics struct {
+	// number of requests matched against specific rule
+	matchedRequests int64
+
+	// number of requests sampled using specific rule
+	sampledRequests int64
+
+	// number of requests borrowed using specific rule
+	borrowedRequests int64
+}
+
+// stale checks if targets (sampling stats) for a given rule is expired or not
+func (r *Rule) stale(now time.Time) bool {
+	reservoirRefreshTime := r.reservoir.refreshedAt.Add(time.Duration(r.reservoir.interval) * time.Second)
+	return r.samplingStatistics.matchedRequests != 0 && now.After(reservoirRefreshTime)
+}
+
+// snapshot takes a snapshot of the sampling statistics counters, returning
+// samplingStatisticsDocument. It also resets statistics counters.
+func (r *Rule) snapshot(now time.Time) *samplingStatisticsDocument {
+	name := r.ruleProperties.RuleName
+
+	matchedRequests := atomic.SwapInt64(&r.samplingStatistics.matchedRequests, int64(0))
+	sampledRequests := atomic.SwapInt64(&r.samplingStatistics.sampledRequests, int64(0))
+	borrowedRequest := atomic.SwapInt64(&r.samplingStatistics.borrowedRequests, int64(0))
+
+	timeStamp := now.Unix()
+	return &samplingStatisticsDocument{
+		RequestCount: &matchedRequests,
+		SampledCount: &sampledRequests,
+		BorrowCount:  &borrowedRequest,
+		RuleName:     &name,
+		Timestamp:    &timeStamp,
+	}
+}
+
+// Sample uses sampling targets of a given rule to decide
+// which sampling should be done and returns a SamplingResult.
+func (r *Rule) Sample(parameters sdktrace.SamplingParameters, now time.Time) sdktrace.SamplingResult {
+	sd := sdktrace.SamplingResult{
+		Tracestate: trace.SpanContextFromContext(parameters.ParentContext).TraceState(),
+	}
+
+	atomic.AddInt64(&r.samplingStatistics.matchedRequests, int64(1))
+
+	// fallback sampling logic if quota for a given rule is expired
+	if r.reservoir.expired(now) {
```

Comment:
How does one get a Manifest that isn't expired that produces a rule with an expired reservoir?  This seems to split the fallback logic into multiple places.

Could we rid ourselves of this fallback code if `Manifest.MatchAgainstManifestRules` also checked if the reservoir (or the Rule) is expired and errored there?

Reply:
manifest update time simply refers to the time when we refreshed all the rules. `reservoir. expired(now)` checks the TTL for a particular rule returned by this [API response](https://docs.aws.amazon.com/xray/latest/api/API_SamplingTargetDocument.html#xray-Type-SamplingTargetDocument-ReservoirQuotaTTL). So non-expired manifest can contain expired rules based on the `reservoirQuotaTTL` value. Yeah I agree we are splitting the fallback logic but we actually need to cover all these cases. Nah I think if rule is expired then we need to sample 1 req/sec and 5% additional request during that second so even if we migrate this to the method you suggested we need this logic there and don't see any benefit of splitting this code since I feel this looks more readable.